Toolkit: simplify the plugin ci docker image

srclosson opened the pull request:
Why this pull request is important:

For build
1) Docker build container was reduced in size (from 5.47GB to 2.88GB)
2) Simplified build by adding shell scripts to do all the hard work
3) Added a testing environment using actual circlci docker container as test bed.

For release process
1) Enterprise plugins now have their source directory removed. Requires having "enterprise: true" in the plugin.json
2) Releases now use nodejs/javascript rather than the ghr tool for a github release. Cleaner, and more manageable.

srclosson committed:
removing src dir on publish

srclosson committed:
removing src dir on publish

srclosson committed:
Moved from binary to native typescript
1. Moved to a native typescrpt github publish using the existing
github client.
2. Change dist.js to detect running in a linked environment.
Todo: Optimize docker image for build size.

srclosson committed:
Moved from binary to native typescript
1. Moved to a native typescrpt github publish using the existing
github client.
2. Change dist.js to detect running in a linked environment.
Todo: Optimize docker image for build size.

srclosson committed:
Optimized build of docker container
Much smaller. From 5.47 gb to 2.88

srclosson committed:
Optimized build of docker container
Much smaller. From 5.47 gb to 2.88

ryantxu commented on packages/grafana-toolkit/docker/grafana-plugin-ci/scripts/deploy.sh:
```diff
@@ -0,0 +1,40 @@
+# We won't validate the sha, as if it's not valid, the deb package won't install, and it will be changing
+# from version to version.
+get_file "https://dl.grafana.com/oss/release/grafana_${version}_amd64.deb" "/tmp/grafana_${version}_amd64.deb"
+dpkg -i "/tmp/grafana_${version}_amd64.deb" && /bin/rm -rf "/tmp/grafana_${version}_amd64.deb"
```
Do we want/need to install it?  I think just download it so it is available with the rest of the bundles.

Is it necessary to have the fancy "latest" detection?  That happens pretty infrequently...

srclosson replied:
Personally, i'd rather leave hooks for interaction that simplify downloading and installing grafana releases.

ryantxu replied:
As discussed -- lets:
* not install any .deb files on the main image
* add a script that lets you install by version.  Support "latest" and "canary" as valid version numbers.  Canary needs to find the last published nightly .deb (some API?)

ryantxu commented on packages/grafana-toolkit/docker/grafana-plugin-ci/Dockerfile:
```diff
@@ -1,51 +1,6 @@
-
-# Download, but don't install previous grafana releases
-RUN mkdir -pv /usr/local/grafana/deb
-ADD https://dl.grafana.com/oss/release/grafana_6.6.2_amd64.deb /usr/local/grafana/deb
-ADD https://dl.grafana.com/oss/release/grafana_6.5.3_amd64.deb /usr/local/grafana/deb
-ADD https://dl.grafana.com/oss/release/grafana_6.4.5_amd64.deb /usr/local/grafana/deb
-ADD https://dl.grafana.com/oss/release/grafana_6.3.7_amd64.deb /usr/local/grafana/deb
```
Is this preserved?  This seems better than having the e2e need to install specific versions

srclosson replied:
No, I'd rather have hooks... I mean, do we keep the last 4 versions? At 80MB a pop, it could be expensive to keep them all. How many do we keep? Each time we release, do we add another? Or just the last 4? 

Are we going to autobuild this VM at release time?

ryantxu replied:
👍

ryantxu commented on packages/grafana-toolkit/src/cli/tasks/plugin.utils.ts:
```diff
@@ -131,28 +138,28 @@ const prepareRelease = useSpinner<any>('Preparing release', async ({ dryrun, ver
 interface GithubPluglishReleaseOptions {
   commitHash?: string;
-  recreate?: boolean;
+  dev?: boolean;
```
what is in dev here?  is this referring to your testing?  is it really dryRun or something?

srclosson replied:
No, if selected, you create a dev branch with a build, rather than a release branch (ie: release-1.4.0-dev", and release a dev release called 1.4.0-dev. Not really specific yet, but more options available as a result of using the rest api.

dprokop commented on packages/grafana-toolkit/bin/grafana-toolkit.dist.js:
```diff
@@ -1,5 +1,22 @@
+entrypoint = () => {
+  const defaultEntryPoint = '../src/cli/index.js';
+  // We are running in dev mode. Don't use compiled binaries, rather use the dev entrypoint.
+  if (fs.existsSync(`${process.env['HOME']}/.config/yarn/link/@grafana/toolkit`)) {
+    console.log('Running in linked mode');
+    return `${__dirname}/grafana-toolkit.js`;
+  }
```
Not sure this is going to work TBH. This file (grafana-toolkit.dist.js) is used by the published package. And in the published tar there is no grafana-toolkit.js file anymore.

Not sure what problem is this `entrypoint` function trying to solve?

srclosson replied:
There were two requirements.
When developing the github publish, I wanted to be able to:
1) Just run `yarn --link @grafana/toolkit` from the destination and nothing more complicated.
2) If running in linked mode, don't use the compiled version, use the typescript version so I can make changes on the fly and test them on the fly.

dprokop commented on packages/grafana-toolkit/src/cli/tasks/plugin.utils.ts:
```diff
@@ -131,28 +138,28 @@ const prepareRelease = useSpinner<any>('Preparing release', async ({ dryrun, ver
 
 interface GithubPluglishReleaseOptions {
```
`Pluglish`, is this an internal joke of Plugins team? :)

srclosson replied:
Did you laugh? Haha, I thought it was pretty funny :) But seriously folks

dprokop replied:
I did! <3

srclosson replied:
Fixed

srclosson committed:
Feedback from discussion with Ryan
- Added gget for getting grafana versions
- Added infrastructure for testing
- Uploaded new docker image

srclosson committed:
Feedback from discussion with Ryan
- Added gget for getting grafana versions
- Added infrastructure for testing
- Uploaded new docker image

srclosson committed:
Fixed typo... Not sure what happened there :)

srclosson committed:
Fixed typo... Not sure what happened there :)

srclosson committed:
Added command to download canary

srclosson committed:
Added command to download canary

srclosson committed:
small fix for displaying versions in help

srclosson committed:
small fix for displaying versions in help

srclosson committed:
Removed --dev option
Should really just rename version to (ex: 1.2.0-dev)

srclosson committed:
Removed --dev option
Should really just rename version to (ex: 1.2.0-dev)

ryantxu reviewed (approved):
Looks good.

srclosson committed:
Merge branch 'remove-src-dir-toolkit' of github.com:grafana/grafana into remove-src-dir-toolkit

srclosson committed:
Merge branch 'master' into remove-src-dir-toolkit
